runtime/middlewares.go

```diff
@@ -0,0 +1,123 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"context"
+
+	jsonschema "github.com/mcuadros/go-jsonschema-generator"
+)
+
+// MiddlewareStack is a stack of Middleware Handlers that can be invoked as an Handler.
+// MiddlewareStack middlewares are evaluated for requests in the order that they are added to the stack
+// followed by the underlying HandlerFn. The middleware responses are then executed in reverse.
+type MiddlewareStack struct {
+	middlewares []MiddlewareHandle
+	handle      HandlerFn
+}
+
+// NewStack returns a new MiddlewareStack instance with no middleware preconfigured.
+func NewStack(middlewares []MiddlewareHandle,
+	handle HandlerFn) *MiddlewareStack {
+	return &MiddlewareStack{
+		handle:      handle,
+		middlewares: middlewares,
+	}
+}
+
+// Middlewares returns a list of all the handlers in the current MiddlewareStack.
+func (m *MiddlewareStack) Middlewares() []MiddlewareHandle {
+	return m.middlewares
+}
+
+// MiddlewareHandle used to define middleware
+type MiddlewareHandle interface {
+	// implement HandleRequest for your middleware.
+	HandleRequest(
+		ctx context.Context,
+		req *ServerHTTPRequest,
+		res *ServerHTTPResponse,
+		shared SharedState) error
+	// implement HandleResponse for your middleware.
+	HandleResponse(
+		ctx context.Context,
+		res *ServerHTTPResponse,
+		shared SharedState) error
+	// return any shared state for this middleware.
+	JSONSchema() *jsonschema.Document
+	Name() string
+}
+
+// SharedState used to access other middlewares in the chain.
+type SharedState struct {
+	middlewareDict map[string]interface{}
+}
+
+func newSharedState(middlewares []MiddlewareHandle) SharedState {
+	sharedState := SharedState{}
+	sharedState.middlewareDict = make(map[string]interface{})
+
+	for i := 0; i < len(middlewares); i++ {
+		sharedState.middlewareDict[middlewares[i].Name()] = middlewares[i]
+	}
+	return sharedState
+}
+
+// GetState returns the state from a different middleware
+func (s SharedState) GetState(name string) interface{} {
+	return s.middlewareDict[name]
+}
+
+// SetState returns the state from a different middleware
+func (s SharedState) SetState(name string, state interface{}) {
+	s.middlewareDict[name] = state
+}
+
+// Handle executes the middlewares in a stack and underlying handler.
+func (m *MiddlewareStack) Handle(
+	ctx context.Context,
+	req *ServerHTTPRequest,
+	res *ServerHTTPResponse) {
+
+	shared := newSharedState(m.middlewares)
+
+	for i := 0; i < len(m.middlewares); i++ {
+		err := m.middlewares[i].HandleRequest(ctx, req, res, shared)
```

Comment:
Add TODO: "A middleware needs to be able to return a bool to say that it has written to the response"

Consider not allowing middlewares to return `error`. When an `error` is returned its ambiguous about who should write to the response.